src/main/java/edu/cornell/kfs/coa/document/validation/impl/GlobalIndirectCostRecoveryAccountsRule.java

```diff
@@ -183,208 +169,193 @@ protected boolean checkIndirectCostRecoveryAccount(IndirectCostRecoveryAccountCh
+		if (totalDistribution.compareTo(ONE_HUNDRED_PERCENT) == 0) {
+			return true;
+		}
+		else{
+			return false;
+		}
+	}
+	
+
+	protected BigDecimal computeTotalDistribution(List<IndirectCostRecoveryAccount> currentActiveIndirectCostRecoveryAccountList) {
+		BigDecimal totalDistribution = BigDecimal.ZERO;
+		if (ObjectUtils.isNotNull(currentActiveIndirectCostRecoveryAccountList) && (currentActiveIndirectCostRecoveryAccountList.size() > 0)) {
+			for (IndirectCostRecoveryAccount icra : currentActiveIndirectCostRecoveryAccountList) {
+				totalDistribution = totalDistribution.add(icra.getAccountLinePercent());
 			}
-			
-			List<IndirectCostRecoveryAccountChange> newICRs = newIcrAccounts;
-			List<IndirectCostRecoveryAccount> updateActive = new ArrayList<IndirectCostRecoveryAccount>();
-			List<IndirectCostRecoveryAccount> addActive = new ArrayList<IndirectCostRecoveryAccount>();
+		}
+		return totalDistribution;
+	}
+	
 
+    protected List<IndirectCostRecoveryAccount> buildCurrentActiveICRAccountsAfterICRUpdates(List<IndirectCostRecoveryAccountChange> icrUpdates, List<IndirectCostRecoveryAccount> existingIcrAccounts, GlobalBusinessObjectDetailBase globalDetail, GlobalObjectWithIndirectCostRecoveryAccounts globalObjectWithIndirectCostRecoveryAccounts){
+    	List<IndirectCostRecoveryAccount> currentActiveIndirectCostRecoveryAccountList = new ArrayList<IndirectCostRecoveryAccount>();				
+		List<IndirectCostRecoveryAccount> existingICRsToBeUpdated = new ArrayList<IndirectCostRecoveryAccount>(); 
+		List<IndirectCostRecoveryAccount> icrAccountsNotUpdated = new ArrayList<IndirectCostRecoveryAccount>();
 
-			if (newIcrAccounts.size() > 0) {
-				for (IndirectCostRecoveryAccountChange newICR : newICRs) {
-					boolean found = false;
-					existingICRs.clear();
-					for (IndirectCostRecoveryAccount icrAccount : existingICRsWorkList) {
-						existingICRs.add(icrAccount);
-					}
-					
-					IndirectCostRecoveryAccount tempIcrAccount = null;
-					int i = 0;
+		for (IndirectCostRecoveryAccount icrAccount : existingIcrAccounts) {
+			icrAccountsNotUpdated.add(icrAccount);
+		}
+		
+		List<IndirectCostRecoveryAccount> icrAccountsToBeUpdatedToActive = new ArrayList<IndirectCostRecoveryAccount>();
+		List<IndirectCostRecoveryAccount> icrAccountsToBeAddedAsActive = new ArrayList<IndirectCostRecoveryAccount>();
+
+
+		if (icrUpdates.size() > 0) {
+			//for each new icr on account global doc search for a matching icr on account to be updated
+			for (IndirectCostRecoveryAccountChange newICR : icrUpdates) {
+				boolean found = false;
+				existingICRsToBeUpdated.clear();
+				for (IndirectCostRecoveryAccount icrAccount : icrAccountsNotUpdated) {
+					existingICRsToBeUpdated.add(icrAccount);
+				}
+				
+				IndirectCostRecoveryAccount tempIcrAccount = null;
+				int i = 0;
+				
+				while(i < existingICRsToBeUpdated.size() && (!found || (found && ObjectUtils.isNotNull(tempIcrAccount))) ){
```

Comment:
Another area where the iterative solution could potentially be replaced with a clearer short-circuiting one. Sorry about suggesting the less-clear iterative strategy in the older review.